src/Microsoft.AspNet.Mvc.Razor/Compilation/RoslynCompilationService.cs

```diff
@@ -300,5 +322,156 @@ private static string ReadFileContentsSafely(IFileProvider fileProvider, string
 
             return null;
         }
+
+        public class WalkerTexasRanger : CSharpSyntaxWalker
+        {
+            public WalkerTexasRanger(SemanticModel semanticModel)
+            {
+                SemanticModel = semanticModel;
+            }
+
+            private SemanticModel SemanticModel { get; }
+
+            public List<SimpleLambdaExpressionSyntax> Expressions { get; } = new List<SimpleLambdaExpressionSyntax>();
+
+            public override void VisitSimpleLambdaExpression(SimpleLambdaExpressionSyntax node)
+            {
+                var type = SemanticModel.GetTypeInfo(node);
+                if (type.ConvertedType.Name != "Expression" && type.ConvertedType.ContainingAssembly.Name != "System.Core")
```

Comment:
This is a prototype. The discussion about the feasibility of the approach + whether or not it adversely affects the dev experience needs to happen first :laughing: 

That said, inside the compiler all you have are names. We should be looking for full name match here, but there isn't a concept of type-equivalence that applies other than matching names. Obvi matching the assembly name doesn't work, I just wanted to bang this out so we could measure the effect.